pkg/deploy/tls.go

```diff
@@ -0,0 +1,448 @@
+//
+// Copyright (c) 2012-2019 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+package deploy
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	stderrors "errors"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	orgv1 "github.com/eclipse/che-operator/pkg/apis/org/v1"
+	"github.com/eclipse/che-operator/pkg/util"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/http/httpproxy"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// TLS related constants
+const (
+	CheTLSJobServiceAccountName           = "che-tls-job-service-account"
+	CheTLSJobRoleName                     = "che-tls-job-role"
+	CheTLSJobRoleBindingName              = "che-tls-job-role-binding"
+	CheTLSJobName                         = "che-tls-job"
+	CheTLSJobComponentName                = "che-create-tls-secret-job"
+	CheTLSSelfSignedCertificateSecretName = "self-signed-certificate"
+	DefaultCheTLSSecretName               = "che-tls"
+)
+
+// IsSelfSignedCertificateUsed detects whether endpoints are/should be secured by self-signed certificate.
+func IsSelfSignedCertificateUsed(checluster *orgv1.CheCluster, clusterAPI ClusterAPI) (bool, error) {
+	if util.IsTestMode() {
+		return true, nil
+	}
+
+	cheTLSSelfSignedCertificateSecret := &corev1.Secret{}
+	err := clusterAPI.Client.Get(context.TODO(), types.NamespacedName{Namespace: checluster.Namespace, Name: CheTLSSelfSignedCertificateSecretName}, cheTLSSelfSignedCertificateSecret)
+	if err == nil {
+		// "self signed-certificate" secret found
+		return true, nil
+	}
+	if !errors.IsNotFound(err) {
+		// Failed to get secret, return error to restart reconcile loop.
+		return false, err
+	}
+
+	if util.IsOpenShift {
+		// Get router TLS certificates chain
+		peerCertificates, err := GetEndpointTLSCrtChain(checluster, "", clusterAPI)
+		if err != nil {
+			return false, err
+		}
+
+		// Check the chain if ti contains self-signed CA certificate
+		for i := range peerCertificates {
```

Comment:
You can avoid using index of element and simple get element from range: `for _, cert := range arr {`

Reply:
done